server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1416,7 +1418,57 @@ public void openEngineAndSkipTranslogRecovery() throws IOException {
         getEngine().skipTranslogRecovery();
     }
 
-    private void innerOpenEngineAndTranslog() throws IOException {
+    /**
+     * Recovers a replica locally up to the global checkpoint before starting a peer recovery.
+     */
+    public void prepareShardForPeerRecovery() {
+        if (state != IndexShardState.RECOVERING) {
```

Comment:
make this an assertion as well same as the other stuff here? Perhaps need safeguard against CLOSE state.
also assert something about `RecoveryState.Stage`?
Should we add a new stage? (will help in cat/recovery)